pyccel/parser/semantic.py

```diff
@@ -464,26 +474,35 @@ def _garbage_collector(self, expr):
         self._allocs.pop()
         return deallocs
 
-    def _infer_type(self, expr, **settings):
+    def _infer_type(self, expr):
         """
-        type inference for expressions
+        Type inference for expressions.
+
+        Create Dictionary of type inference for data type
+        passed in expr.
+
+        Parameters
+        ----------
+        expr : data-type
+                Data type.
+
+        Returns
+        -------
+        d_var
+            Dictionary of type inference for expr.
```

Comment:
The type of the parameter is not accurate here, and I don't think "type inference" is very well defined. What about:
```suggestion
        Infer all relevant type information for the expression.

        Create a dictionary describing all the type information that can be
        inferred about the expression `expr`. This includes information about:
        - `datatype`
        - `precision`
        - `rank`
        - `shape`
        - `order`
        - `memory_handling`
        - `cls_base`
        - `is_target`

        Parameters
        ----------
        expr : pyccel.ast.basic.Basic
                An AST object representing an object in the code whose type
                must be determined.

        Returns
        -------
        dict
            Dictionary containing all the type information which was inferred.
```